draft-specifications/Model.md

```diff
@@ -0,0 +1,326 @@
+
+This document captures a summary of consensus and remaining contention for the
+OCapN data model and abstract syntax, excluding the concern of concrete
+representation of these on the wire, but including non-normative
+representations in a selection of implementation languages.
+
+Commentary in block quotes is not normative.
+
+> Non-normative commentary.
+
+# Atom
+
+## Undefined
+
+A value representing the absence of a value.
+
+> - **Guile**: `*unspecified*`
+> - **JavaScript**: `undefined`
+> - **Python**: `None`
+
+## Null (JSON)
+
+A value representing `null` as distinct from `undefined` in JSON.
+
+> In order for the JSON subset of JavaScript to be passable, Undefined must be
+> distinguishable from Null.
```

Comment:
Hmm, this reads strangely. Isn't JSON inherently passable?

Reply:
There are caveats.

- JSON documents that contain any string that includes a Unicode surrogate is not passable.
- JSON documents with numbers that exceed the precision available to an IEEE 64 bit floating point number are not passable.

The second caveat isn’t really a caveat for the ECMA 404 version of JSON but the Unicode limitation does not admit all JSON expressible in JavaScript.

This comment speaks more to the point that if OCapN didn’t have separate Null and Undefined, JSON documents that had an explicit `null` in them would not have pass-invariant structural equivalence (be “passable”). Null and undefined would have to be collapsed into a single type, such that either struct fields with null values were omitted when they returned, or struct fields will undefined values would return as null.

Reply:
I’ve posted a commit that avoids the jargon “passable” and makes painfully clear the scenario that OCapN holds invariant by distinguishing Null from Undefined.

Reply:
> * JSON documents that contain any string that includes a Unicode surrogate is not passable.
> * JSON documents with numbers that exceed the precision available to an IEEE 64 bit floating point number are not passable.

Annals of "everything is exactly the same": [RFC 7493 I-JSON](https://www.rfc-editor.org/rfc/rfc7493)
> Object member names, and string values in arrays and object members, MUST NOT include code points that identify Surrogates or Noncharacters as defined by [<a href="https://www.rfc-editor.org/rfc/rfc7493#ref-UNICODE" title="&quot;The Unicode Standard&quot;">UNICODE</a>].
> …
> I-JSON messages SHOULD NOT include numbers that express greater magnitude or precision than an IEEE 754 double precision number provides, for example, 1E400 or 3.141592653589793238462643383279.
> …
> An I-JSON sender cannot expect a receiver to treat an integer whose absolute value is greater than 9007199254740991 (i.e., that is outside the range [-(2\*\*53)+1, (2\*\*53)-1]) as an exact value.
> …
> Objects in I-JSON messages MUST NOT have members with duplicate names.